### DISCUSSION TITLE
How to create a thread safe logger with multiple sinks?

### DISCUSSION THREAD

**Discussion by Linhuihang** [gabime/spdlog#2493] [Q&A] [answered]
I know we should use factory method to create a thread safe logger, just like: spdlog::basic_logger_mt or spdlog::rotating_logger_mt.
But it seems we can only use factory method to create logger with only one sink?
So my question is how to create a thread safe logger with multiple sinks?
Just like: spdlog::logger logger("multi_sink", { consoleSink, rotatingLogger });

**Comment by tt4g** [comment, marked as answer]
> So my question is how to create a thread safe logger with multiple sinks?

The logger with multiple sinks is supported.
See Wiki: https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-loggers-with-multiple-sinks

**Reply by Linhuihang** [reply]
But this is not the factory method. So the logger created in this way is not thread safe.
( According to https://spdlog.docsforge.com/v1.x/1.1.thread-safety/, we should use the _mt factory functions to create thread safe loggers. )

**Reply by tt4g** [reply]
> Note: This restriction applies to all kind of loggers ("_mt" or "_st").
>
> https://github.com/gabime/spdlog/wiki/1.1.-Thread-Safety

This sentence in the Wiki is strictly incorrect.

The factory method simply calls `spdlog::synchronous_factory::create`.

https://github.com/gabime/spdlog/blob/81de01c02c2b49aa723806b0c1d557dd05721635/include/spdlog/sinks/rotating_file_sink.h#L62-L68

The `spdlog::synchronous_factory::create` just creates the logger and sink.

https://github.com/gabime/spdlog/blob/81de01c02c2b49aa723806b0c1d557dd05721635/include/spdlog/details/synchronous_factory.h#L13-L23

In other words, the logger remains at the same thread-safe level whether the factory method is used or not.
The thread-safe guarantee of sink depends on the sink name prefix (`_mt` and `_st`).

**Reply by Linhuihang** [reply]
Thanks for your reply.
